EducationPortal/EducationPortal.BLL/Services/CourseService.cs

```diff
@@ -0,0 +1,297 @@
+        private IRepository<Skill> skills;
+        private IRepository<Material> materials;
+
+        private Mapper courseMapper;
+        private Mapper skillMapper;
+
+        public string Name => "Course";
+        
+        public CourseService(IRepository<Course> courses, 
+                             IRepository<Skill> skills,
+                             IRepository<Material> materials)
+        {
+            this.courses = courses;
+            this.skills = skills;
+            this.materials = materials;
+
+            var skillConfig = new MapperConfiguration(cfg => cfg.CreateMap<SkillDTO, Skill>().ReverseMap());
+            this.skillMapper = new Mapper(skillConfig);
+
+            var courseConfig = new MapperConfiguration(cfg => 
+            {
+                cfg.CreateMap<CourseDTO, Course>()
+                .ForMember(dest => dest.SkillIds,
+                           opt => opt.MapFrom(dto => skills.Find(skill => skill.Name == dto.Name).Select(skill => skill.Id).ToArray()));
+
+                cfg.CreateMap<Course, CourseDTO>()
+                .ForMember(dest => dest.Skills, 
+                           opt => opt.MapFrom(course => skillMapper.Map<IEnumerable<SkillDTO>>(skills.Find(a => course.SkillIds.Contains(a.Id)))));
+            });
+            this.courseMapper = new Mapper(courseConfig);
+        }
+
+        public OperationResponse AddCourse(CourseDTO course)
+        {
+            var response = new OperationResponse();
+            var courseToAdd = courseMapper.Map<Course>(course);
+            courseToAdd.MaterialIds = new long[0];
+            courseToAdd.SkillIds = new long[0];
+            courses.Create(courseToAdd);
+            courses.Save();
+            response.Message = "Новый курс успешно создан!";
+            return response;
+        }
+
+        public GetCoursesResponse GetUserCourses(long userId)
+        {
+            var response = new GetCoursesResponse();
+
+            var userCourses = courses.Find(course => course.CreatorId == userId);
+
+            response.Courses = courseMapper.Map<IEnumerable<Course>, IEnumerable<CourseDTO>>(userCourses);
+
+            return response;
+        }
+
+        public GetCoursesResponse GetAllCourses()
+        {
+            var response = new GetCoursesResponse();
+
+            var userCourses = courses.GetAll();
+
+            response.Courses = courseMapper.Map<IEnumerable<Course>, IEnumerable<CourseDTO>>(userCourses);
+
+            return response;
+        }
+
+        public OperationResponse EditCourse(long userId, CourseDTO newCourseInfo)
+        {
+            var response = new OperationResponse();
+
+            var courseToUpdate = courses.Find(course => course.CreatorId == userId
+                                                     && newCourseInfo.Id == course.Id).SingleOrDefault();
+
+            if (courseToUpdate == null)
+            {
+                response.Message = "Вы не являетесь автором данного курса";
```

Comment:
Is this a correct error?

Reply:
Use constants for messages.

Reply:
> Is this a correct error?

Any user can type "editcourse" command in console version. That's why we need this check. I have already replaced it by "CanEditCourse" method call.